File: backend/app/models/rbac.py
# backend/app/models/rbac.py
"""
RBAC Data Models Module

This module defines Pydantic models for representing Kubernetes RBAC (Role-Based Access Control)
data structures. These models are used for API request/response serialization and validation
throughout the Kube-Guard application.

The models cover:
- RBAC policy rules and permissions
- Security findings and recommendations  
- Subject and role binding relationships
- Kubernetes resource access patterns

Author: Gerardo Zapico
Date: July 2025
"""

from pydantic import BaseModel
from typing import Optional, List

class RbacOrigin(BaseModel):
    """
    Represents the origin source of an RBAC permission.
    
    Attributes:
        apiGroup (str): Kubernetes API group (e.g., 'apps', 'core')
        kind (str): Resource kind (e.g., 'Role', 'ClusterRole')
        name (str): Name of the RBAC resource
    """
    apiGroup: str
    kind: str
    name: str

class RbacAllowedAction(BaseModel):
    """
    Represents a specific action that is allowed by RBAC policies.
    
    Attributes:
        namespace (str): Target namespace for the action
        verb (str): Kubernetes verb (get, list, create, delete, etc.)
        apiGroup (Optional[str]): API group of the target resource
        resource (Optional[str]): Resource type (pods, services, etc.)
        nonResourceURI (Optional[str]): Non-resource URL path
        originatedFrom (List[RbacOrigin]): List of RBAC sources granting this permission
    """
    namespace: str
    verb: str
    apiGroup: Optional[str] = None
    resource: Optional[str] = None
    nonResourceURI: Optional[str] = None
    originatedFrom: List[RbacOrigin]

class Subject(BaseModel):
    """
    Represents a Kubernetes RBAC subject (User, Group, or ServiceAccount).
    
    Attributes:
        kind (str): Type of subject ('User', 'Group', 'ServiceAccount')
        name (str): Name of the subject
        namespace (Optional[str]): Namespace (required for ServiceAccounts)
        apiGroup (Optional[str]): API group for the subject
    """
    kind: str
    name: str
    namespace: Optional[str] = None
    apiGroup: Optional[str] = None

class FindingDetails(BaseModel):
    """
    Detailed information about a security finding.
    
    Attributes:
        Severity (str): Severity level ('High', 'Medium', 'Low')
        Message (str): Human-readable description of the finding
        Recommendation (str): Suggested remediation steps
        RuleName (str): Name of the security rule that triggered
        RuleUuid (str): Unique identifier for the rule
        References (Optional[List[str]]): Related documentation or CVE references
    """
    Severity: str
    Message: str
    Recommendation: str
    RuleName: str
    RuleUuid: str
    References: Optional[List[str]] = []

class RbacFinding(BaseModel):
    """
    Represents a security finding related to RBAC configuration.
    
    Attributes:
        Subject (Subject): The subject involved in the finding
        Finding (FindingDetails): Details about the security issue
    """
    Subject: Subject
    Finding: FindingDetails

class RbacFindingWithRules(RbacFinding):
    """
    Extended RBAC finding that includes detailed policy rules and context.
    
    Inherits from RbacFinding and adds:
        AllowedActions (List[RbacAllowedAction]): Actions permitted by the policy
        PolicyRules (List[RbacPolicyRule]): Detailed policy rules
        Namespace (Optional[str]): Target namespace
        Resource (Optional[str]): Target resource type
        ResourceName (Optional[str]): Specific resource name
        NonResourceURL (Optional[str]): Non-resource URL
        Verb (Optional[str]): Specific action verb
    """
    AllowedActions: List[RbacAllowedAction]
    PolicyRules: List['RbacPolicyRule']  
    Namespace: Optional[str] = None
    Resource: Optional[str] = None
    ResourceName: Optional[str] = None
    NonResourceURL: Optional[str] = None
    Verb: Optional[str] = None
    
class RbacPolicyRule(BaseModel):
    """
    Represents a policy rule with its allowed actions.
    
    Attributes:
        kind (str): Type of policy ('Role' or 'ClusterRole')
        name (str): Name of the policy
        namespace (Optional[str]): Namespace (for Roles only)
        allowedTo (List[RbacAllowedAction]): List of permitted actions
    """
    kind: str
    name: str
    namespace: Optional[str] = None
    allowedTo: List[RbacAllowedAction]

class K8sPolicyRule(BaseModel):
    """
    Represents a Kubernetes native policy rule structure.
    
    This mirrors the structure of PolicyRule from the Kubernetes API,
    defining what actions are allowed on which resources.
    
    Attributes:
        verbs (List[str]): List of allowed verbs (get, list, create, etc.)
        apiGroups (Optional[List[str]]): API groups the rule applies to
        resources (Optional[List[str]]): Resource types the rule applies to
        resourceNames (Optional[List[str]]): Specific resource names
        nonResourceURLs (Optional[List[str]]): Non-resource URL paths
    """
    verbs: List[str]
    apiGroups: Optional[List[str]] = []
    resources: Optional[List[str]] = []
    resourceNames: Optional[List[str]] = []
    nonResourceURLs: Optional[List[str]] = []

class RbacSubject(BaseModel):
    """
    Represents a subject in an RBAC binding.
    
    Attributes:
        kind (str): Subject type ('User', 'Group', 'ServiceAccount')
        name (str): Subject name
        apiGroup (Optional[str]): API group (typically empty for core subjects)
        namespace (Optional[str]): Namespace for ServiceAccounts
    """
    kind: str
    name: str
    apiGroup: Optional[str] = ""
    namespace: Optional[str] = None  # Namespace is optional for subjects

class RbacRoleRef(BaseModel):
    """
    Represents a reference to a Role or ClusterRole in a binding.
    
    Attributes:
        kind (str): Type of role ('Role' or 'ClusterRole')
        name (str): Name of the role
        apiGroup (Optional[str]): API group (typically 'rbac.authorization.k8s.io')
        namespace (Optional[str]): Namespace (for Role references)
    """
    kind: str
    name: str
    apiGroup: Optional[str] = ""
    namespace: Optional[str] = None  # Namespace is optional for role references

class RbacBinding(BaseModel):
    """
    Represents a complete RBAC binding (RoleBinding or ClusterRoleBinding).
    
    Attributes:
        id (int): Unique identifier for the binding
        name (str): Name of the binding
        kind (str): Type of binding ('RoleBinding' or 'ClusterRoleBinding')
        subjects (List[RbacSubject]): List of subjects bound to the role
        roleRef (RbacRoleRef): Reference to the role being bound
        raw (Optional[str]): Raw YAML representation of the binding
    """
    id: int
    name: str
    kind: str  # RoleBinding o ClusterRoleBinding
    subjects: List[RbacSubject]
    roleRef: RbacRoleRef
    raw: Optional[str] = None  # YAML raw data

File: frontend/kube-guard-frontend/src/types/rbac.ts
/**
 * RBAC Type Definitions
 * 
 * This module contains TypeScript type definitions for all RBAC-related data
 * structures used throughout the KubeGuard frontend application. These types
 * ensure type safety and provide clear contracts for API responses and component props.
 * 
 * The types mirror the backend API responses and provide a consistent interface
 * for working with Kubernetes RBAC data including bindings, findings, and policy rules.
 * 
 * @author Gerardo Zapico
 * @version 1.0.0
 */

/**
 * Represents a subject in an RBAC binding.
 * 
 * Subjects are entities (users, service accounts, groups) that can be granted
 * permissions through RBAC bindings.
 */
export interface RbacSubject {
  /** The kind of subject (User, ServiceAccount, Group) */
  kind: string
  
  /** The name of the subject */
  name: string
  
  /** API group for the subject (optional) */
  apiGroup?: string | null
  
  /** Namespace for the subject (optional, used for ServiceAccounts) */
  namespace?: string | null
}

/**
 * Represents a role reference in an RBAC binding.
 * 
 * Role references point to the Role or ClusterRole that defines the permissions
 * being granted to the subjects in the binding.
 */
export interface RbacRoleRef {
  /** The kind of role (Role or ClusterRole) */
  kind: string
  
  /** The name of the role */
  name: string
  
  /** API group for the role (optional) */
  apiGroup?: string | null
  
  /** Namespace for the role (required for Roles, null for ClusterRoles) */
  namespace?: string | null
}

/**
 * Represents a complete RBAC binding (RoleBinding or ClusterRoleBinding).
 * 
 * Bindings connect subjects to roles, granting the permissions defined in
 * the role to the specified subjects.
 */
export interface RbacBinding {
  /** Unique identifier for the binding */
  id: number
  
  /** Name of the binding */
  name: string
  
  /** Kind of binding (RoleBinding or ClusterRoleBinding) */
  kind: string
  
  /** Array of subjects that receive the permissions */
  subjects: RbacSubject[]
  
  /** Reference to the role that defines the permissions */
  roleRef: RbacRoleRef
  
  /** Raw YAML representation of the binding (optional) */
  raw?: string
}

/**
 * Legacy subject interface for compatibility.
 * @deprecated Use RbacSubject instead
 */
export interface Subject {
  kind: string
  name: string
  namespace?: string
  apiGroup?: string
}

/**
 * Detailed information about a security finding.
 * 
 * Contains the analysis results including severity, recommendations,
 * and references for addressing the security issue.
 */
export interface FindingDetails {
  /** Severity level of the finding (High, Medium, Low) */
  Severity: string
  
  /** Descriptive message explaining the finding */
  Message: string
  
  /** Recommended action to address the finding */
  Recommendation: string
  
  /** Name of the rule that generated this finding */
  RuleName: string
  
  /** Unique identifier for the rule */
  RuleUuid: string
  
  /** Optional references for additional information */
  References?: string[]
}

/**
 * Represents a security finding from RBAC analysis.
 * 
 * Findings identify potential security risks or policy violations
 * in the cluster's RBAC configuration.
 */
export interface RbacFinding {
  /** The subject associated with this finding */
  Subject: Subject
  
  /** Detailed information about the finding */
  Finding: FindingDetails
}

/**
 * Represents the origin of a permission (which role granted it).
 * 
 * Used to track where specific permissions come from in the RBAC hierarchy.
 */
export interface RbacOrigin {
  /** API group of the role that grants this permission */
  apiGroup: string
  
  /** Kind of the role (Role or ClusterRole) */
  kind: string
  
  /** Name of the role */
  /** Name of the role */
  name: string
}

/**
 * Represents an allowed action/permission that a subject can perform.
 * 
 * Allowed actions define specific permissions including the verb (action),
 * target resources, and the roles that grant these permissions.
 */
export interface RbacAllowedAction {
  /** Namespace where this action is allowed */
  namespace: string
  
  /** Action verb (get, list, create, update, delete, etc.) */
  verb: string
  
  /** API group of the target resource (optional) */
  apiGroup?: string
  
  /** Type of resource this action applies to (optional) */
  resource?: string
  
  /** Non-resource URI for actions outside of resource types (optional) */
  nonResourceURI?: string
  
  /** Array of roles/origins that grant this permission */
  originatedFrom: RbacOrigin[]
}

/**
 * Represents a complete policy rule for a subject.
 * 
 * Policy rules aggregate all permissions that a specific subject has,
 * providing a comprehensive view of what actions they can perform.
 */
export interface RbacPolicyRule {
  /** Kind of subject this rule applies to */
  kind: string
  
  /** Name of the subject */
  name: string
  
  /** Namespace of the subject (optional) */
  namespace?: string
  
  /** Array of all actions this subject is allowed to perform */
  allowedTo: RbacAllowedAction[]
}

/**
 * Represents a Kubernetes native policy rule structure.
 * 
 * This mirrors the structure of PolicyRule from the Kubernetes API,
 * defining what actions are allowed on which resources for a specific Role or ClusterRole.
 */
export interface K8sPolicyRule {
  /** List of allowed verbs (get, list, create, update, patch, delete, etc.) */
  verbs: string[]
  
  /** API groups the rule applies to (optional) */
  apiGroups?: string[]
  
  /** Resource types the rule applies to (optional) */
  resources?: string[]
  
  /** Specific resource names (optional) */
  resourceNames?: string[]
  
  /** Non-resource URL paths (optional) */
  nonResourceURLs?: string[]
}


